mongodb/collection.go

```diff
@@ -47,9 +47,62 @@ func (c *Collection) Must() *Must {
 	return newMust(c)
 }
 
-// Find returns a Find interface which can be used to either refine the criteria or retrieve a cursor
-func (c *Collection) Find(query interface{}) *Find {
-	return newFind(c.collection, query)
+// Distinct returns the list of distinct values for the given field name in the collection
+func (c *Collection) Distinct(ctx context.Context, filter interface{}, fieldName string) ([]interface{}, error) {
+
+	results, err := c.collection.Distinct(ctx, fieldName, filter)
+
+	return results, wrapMongoError(err)
+}
+
+// Count returns the number of documents in the collection that satisfy the given filter (which cannot be nil)
+// Sort and Projection options are ignored. A Limit option <=0 is ignored, and a count of all documents is returned
+func (c *Collection) Count(ctx context.Context, filter interface{}, opts ...FindOption) (int, error) {
+
+	count, err := c.collection.CountDocuments(ctx, filter, newFindOptions(opts...).asDriverCountOption())
+
+	return int(count), wrapMongoError(err)
+}
+
+// Find returns the total number of documents in the collection that satisfy the given filter (restricted by the
+// given options), with the actual documents provided in the results parameter (which must be a non nil pointer
+// to a slice of the expected document type)
+// If no sort order option is provided a default sort order of 'ascending _id' is used (bson.M{"_id": 1})
+func (c *Collection) Find(ctx context.Context, filter, results interface{}, opts ...FindOption) (int, error) {
+
+	fo := newFindOptions(opts...)
+
+	tc, err := c.collection.CountDocuments(ctx, filter)
+	switch {
+	case err != nil:
+		return 0, err
+	case tc == 0:
+		return 0, nil
+	case fo.limit < 0 || fo.skip >= tc,
+		fo.limit == 0 && fo.obeyZeroLimit:
```

Comment:
I find it a bit confusing using multiple conditions like this when they are all logical operations. Maybe have a single condition?
```suggestion
	case fo.limit < 0 || fo.skip >= tc || (fo.limit == 0 && fo.obeyZeroLimit):
```

Reply:
This one I'd push back on - I find it much more readable (and consequently less confusing) to have the multiple case statements, and it's very idiomatic in go (although you are right of course that they are logical or's)

Reply:
Fair enough. But in that case maybe use 3 conditions instead?

```suggestion
	case fo.limit < 0, fo.skip >= tc,
		fo.limit == 0 && fo.obeyZeroLimit:
```

Reply:
Ok I see it is actually 2 conditions but one of them happens to be an OR. Agreed to leave it like this